pkg/authorization/authorizer/scope/converter.go

```diff
@@ -0,0 +1,141 @@
+package scope
+
+import (
+	"fmt"
+	"strings"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kapierrors "k8s.io/kubernetes/pkg/api/errors"
+	kutilerrors "k8s.io/kubernetes/pkg/util/errors"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	"github.com/openshift/origin/pkg/authorization/rulevalidation"
+	userapi "github.com/openshift/origin/pkg/user/api"
+)
+
+// ScopesToRules takes the scopes and return the rules back.  We ALWAYS add the discovery rules and it is possible to get some rules and and
+// an error since errors aren't fatal to evaluation
+func ScopesToRules(scopes []string, namespace string, clusterPolicyGetter rulevalidation.ClusterPolicyGetter) ([]authorizationapi.PolicyRule, error) {
+	rules := append([]authorizationapi.PolicyRule{}, discoveryRules...)
+
+	errors := []error{}
+	for _, scope := range scopes {
+		found := false
+
+		for prefix, evaluator := range scopeEvaluators {
+			if strings.HasPrefix(scope, prefix) {
+				found = true
+				currRules, err := evaluator(scope, namespace, clusterPolicyGetter)
+				if err != nil {
+					errors = append(errors, err)
+					continue
+				}
+
+				rules = append(rules, currRules...)
+			}
+		}
+
+		if !found {
+			errors = append(errors, fmt.Errorf("no scope evaluator found for %q", scope))
+		}
+	}
+
+	return rules, kutilerrors.NewAggregate(errors)
+}
+
+var discoveryRules = []authorizationapi.PolicyRule{
+	{
+		Verbs: sets.NewString("get"),
+		NonResourceURLs: sets.NewString(
+			// API discovery/negotiation
+			"/api", "/api/*",
+			"/apis", "/apis/*",
+			"/oapi", "/oapi/*",
+		),
+	},
+}
+
+const (
+	SpecialIndicator       = "special:"
+	ClusterRoleIndicator   = "clusterrole:"
+	ClusterWideIndicator   = "clusterwide:"
+	NamespaceWideIndicator = "namespace:"
+)
+
+// scopeEvaluator takes a scope and returns the rules that express it
+type scopeEvaluator func(scope, namespace string, clusterPolicyGetter rulevalidation.ClusterPolicyGetter) ([]authorizationapi.PolicyRule, error)
+
+// scopeEvaluators map prefixes to a function that handles that prefix
+var scopeEvaluators = map[string]scopeEvaluator{
+	SpecialIndicator:     specialEvaluator,
+	ClusterRoleIndicator: clusterRoleEvaluator,
```

Comment:
> rather than a prefix:evaluator map, I'd prefer a list of evaluators, and make one of the return values whether the evaluator recognized/evaluated the scope, and short-circuit the list when one recognizes the scope format. I'd like to avoid the "special:" prefix if possible

I've regretted not having a map in the past (think our resource to handler maps in some admission plugins), but I don't feel strongly about that.

I do really want a "special" list for curated options instead of making roles to satisfy them.  Having a curated list for the standard or most common options is easier to describe than a role which is mutable.  In addition, while roles may seem nice, I see enumerated permissions as more reliable since roles are at the mercy of the cluster-admin and may not even exist in all clusters.  I think that roles are what people think they want, but in terms of portability they will not be particularly reliable for contained units.

Short version, "special" exists to provide a guaranteed way of saying "exactly this set of permissions, no matter how this cluster is configured" and proof of identity and can I do this are both important enough to warrant inclusion.